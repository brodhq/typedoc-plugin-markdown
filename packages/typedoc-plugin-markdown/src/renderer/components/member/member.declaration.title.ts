import {
  DeclarationReflection,
  ParameterReflection,
  ReflectionKind,
} from 'typedoc';
import { LiteralType, ReflectionType } from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { escapeChars, stripComments, stripLineBreaks } from '../../utils';
import { TypeComponent } from '../shared/type';
import { MemberSymbolComponent } from './member.symbol';

export function DeclarationTitleComponent(
  model: ParameterReflection | DeclarationReflection,
) {
  const md = new MarkdownBuilder();

  md.add(MemberSymbolComponent(model));

  if (model.flags && model.flags.length > 0 && !model.flags.isRest) {
    md.add(' ' + model.flags.map((flag) => `\`${flag}\``).join(' '));
  }
  md.add(`${model.flags.isRest ? '... ' : ''} **${escapeChars(model.name)}**`);
  if (model instanceof DeclarationReflection && model.typeParameters) {
    md.add(
      `<${model.typeParameters
        .map((typeParameter) => `\`${typeParameter.name}\``)
        .join(', ')}\\>`,
    );
  }

  if (!model.parent?.kindOf(ReflectionKind.Enum)) {
    md.add(getType(model));
  }

  if (
    !(model.type instanceof LiteralType) &&
    model.defaultValue &&
    model.defaultValue !== '...'
  ) {
    md.add(
      ` = \`${stripLineBreaks(
        escapeChars(stripComments(model.defaultValue)),
      )}\``,
    );
  }
  return md.print('');
}

function getType(reflection: ParameterReflection | DeclarationReflection) {
  const reflectionType = reflection.type as ReflectionType;
  if (reflectionType && reflectionType.declaration?.children) {
    return ': `Object`';
  }
  return (
    ': ' + TypeComponent(reflectionType ? reflectionType : reflection, 'object')
  );
}
