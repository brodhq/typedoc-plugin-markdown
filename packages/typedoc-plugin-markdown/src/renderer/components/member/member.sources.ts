import { DeclarationReflection, SignatureReflection } from 'typedoc';
import { ArrayType, ReferenceType, Type } from 'typedoc/dist/lib/models/types';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { heading, link } from '../../../markdown-tools/elements';
import { escapeChars } from '../../utils';
import { LinkComponent } from '../shared/link';

export function SourcesComponent(
  model: DeclarationReflection | SignatureReflection,
) {
  const md = new MarkdownBuilder();

  if (model.implementationOf) {
    md.add(heading(4, 'Implementation of'));
    md.add(TypeAndParentComponent(model.implementationOf));
  }

  if (model.inheritedFrom) {
    md.add(heading(4, 'Inherited from'));
    md.add(TypeAndParentComponent(model.inheritedFrom));
  }

  if (model.overwrites) {
    md.add(heading(4, 'Overrides'));
    md.add(TypeAndParentComponent(model.overwrites));
  }

  if (model.sources) {
    md.add(heading(4, 'Defined in'));

    model.sources.forEach((source) => {
      if (source.url) {
        md.add(link(`${source.fileName}:${source.line}`, source.url));
      } else {
        md.add(`${source.fileName}:${source.line}`);
      }
    });
  }

  return md.print();
}

export function TypeAndParentComponent(
  model: ArrayType | ReferenceType | Type,
) {
  if (model) {
    if ('elementType' in model) {
      return TypeAndParentComponent(model.elementType) + '[]';
    } else {
      if (model instanceof ReferenceType && model.reflection) {
        const md: string[] = [];
        if (model.reflection instanceof SignatureReflection) {
          if (model.reflection.parent?.parent?.url) {
            md.push(
              getUrl(
                model.reflection.parent.parent.name,
                model.reflection.parent.parent.url,
              ),
            );
            if (model.reflection.parent.url) {
              md.push(
                getUrl(
                  model.reflection.parent.name,
                  model.reflection.parent.url,
                ),
              );
            }
          }
        } else {
          if (model.reflection.parent?.url) {
            md.push(
              getUrl(model.reflection.parent.name, model.reflection.parent.url),
            );
            if (model.reflection.url) {
              md.push(getUrl(model.reflection.name, model.reflection.url));
            }
          }
        }
        return md.join('.');
      } else {
        return escapeChars(model.toString());
      }
    }
  }
  return 'void';
}

const getUrl = (name: string, url: string) => LinkComponent(name, url);
