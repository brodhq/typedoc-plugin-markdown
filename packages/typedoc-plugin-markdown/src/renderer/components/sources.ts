import { DeclarationReflection, SignatureReflection } from 'typedoc';

import { MarkdownBuilder } from '../../markdown-tools//builder';
import { heading, link } from '../../markdown-tools//elements';
import { TypeAndParentComponent } from './type-and-parent';

export function SourcesComponent(
  model: DeclarationReflection | SignatureReflection,
) {
  const md = new MarkdownBuilder();

  if (model.implementationOf) {
    md.add(heading(4, 'Implementation of'));
    md.add(TypeAndParentComponent(model.implementationOf));
  }

  if (model.inheritedFrom) {
    md.add(heading(4, 'Inherited from'));
    md.add(TypeAndParentComponent(model.inheritedFrom));
  }

  if (model.overwrites) {
    md.add(heading(4, 'Overrides'));
    md.add(TypeAndParentComponent(model.overwrites));
  }

  if (model.sources) {
    md.add(heading(4, 'Defined in'));

    model.sources.forEach((source) => {
      if (source.url) {
        md.add(link(`${source.fileName}:${source.line}`, source.url));
      } else {
        md.add(`${source.fileName}:${source.line}`);
      }
    });
  }

  return md.print();
}
