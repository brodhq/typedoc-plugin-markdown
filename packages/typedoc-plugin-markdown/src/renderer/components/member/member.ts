import { DeclarationReflection, ReferenceReflection } from 'typedoc';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { heading } from '../../../markdown-tools/elements';
import { useState } from '../../store';
import { escapeChars } from '../../utils';
import { DeclarationComponent } from './member.declaration';
import { SignatureComponent } from './member.signature';

export function MemberComponent(
  model: DeclarationReflection | ReferenceReflection,
) {
  const { options } = useState();
  const md = new MarkdownBuilder();

  md.add(
    heading(
      3,
      options.namedAnchors
        ? `<a id="${model.anchor}" name="${model.anchor}"></a> `
        : '' + escapeChars(model.name),
    ),
  );

  if (model.signatures) {
    model.signatures.forEach((signature) => {
      md.add(SignatureComponent(signature));
    });
  } else {
    if (model.getSignature) {
      md.add(SignatureComponent(model.getSignature));
    }

    if (model.setSignature) {
      md.add(SignatureComponent(model.setSignature));
    }

    if (model instanceof ReferenceReflection && model.isReference) {
      const deep = model.tryGetTargetReflectionDeep();
      if (deep) {
        md.add(`Re-exports ${deep.name}`);
      }
    }

    if (model instanceof DeclarationReflection) {
      md.add(DeclarationComponent(model));
    }
  }

  return md.print();
}
