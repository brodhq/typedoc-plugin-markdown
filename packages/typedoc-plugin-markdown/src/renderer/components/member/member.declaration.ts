import {
  DeclarationReflection,
  SignatureReflection,
} from 'typedoc/dist/lib/models';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { heading } from '../../../markdown-tools/elements';
import { CommentsComponent } from '../shared/comments';
import { PropertyTableComponent } from '../shared/table.property';
import { TypeParameterTableComponent } from '../shared/table.type-parameter';
import { TypeComponent } from '../shared/type';
import { DeclarationTitleComponent } from './member.declaration.title';
import { SignatureComponent } from './member.signature';
import { SourcesComponent } from './member.sources';

export function DeclarationComponent(model: DeclarationReflection) {
  const md = new MarkdownBuilder();

  const typeDeclaration = (model.type as any)
    ?.declaration as DeclarationReflection;

  md.add(DeclarationTitleComponent(model));

  if (model.comment) {
    md.add(CommentsComponent(model.comment));
  }

  if (model.typeParameters) {
    md.add(heading(4, 'Type parameters'));
    md.add(TypeParameterTableComponent(model.typeParameters));
  }

  if (typeDeclaration?.indexSignature) {
    md.add(heading(4, 'Index signature'));
    md.add(getIndexSignatureTitle(typeDeclaration.indexSignature));
  }

  if (typeDeclaration?.signatures) {
    md.add(
      heading(
        4,
        typeDeclaration.children ? 'Call signature' : 'Type declaration',
      ),
    );
    typeDeclaration.signatures.forEach((signature) => {
      md.add(SignatureComponent(signature, true));
    });
  }

  if (typeDeclaration?.children) {
    md.add(heading(4, 'Type declaration'));
    md.add(PropertyTableComponent(typeDeclaration.children));
  }

  md.add(SourcesComponent(model));

  return md.print();
}

export function getIndexSignatureTitle(model: SignatureReflection) {
  const md = ['â–ª'];
  const parameters = model.parameters
    ? model.parameters.map((parameter) => {
        return `${parameter.name}: ${
          parameter.type ? TypeComponent(parameter.type) : ''
        }`;
      })
    : [];
  md.push(
    `\[${parameters.join('')}\]: ${
      model.type ? TypeComponent(model.type) : ''
    }`,
  );
  return md.join(' ');
}
