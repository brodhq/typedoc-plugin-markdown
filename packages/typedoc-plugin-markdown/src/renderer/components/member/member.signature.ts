import {
  DeclarationReflection,
  ReflectionKind,
  SignatureReflection,
} from 'typedoc';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { heading } from '../../../markdown-tools/elements';
import { CommentsComponent, comment } from '../shared/comments';
import { ParameterTable } from '../shared/table.parameter';
import { PropertyTableComponent } from '../shared/table.property';
import { TypeParameterTableComponent } from '../shared/table.type-parameter';
import { TypeComponent } from '../shared/type';
import { SignatureTitleComponent } from './member.signature.title';
import { SourcesComponent } from './member.sources';

export function SignatureComponent(model: SignatureReflection, nested = false) {
  const md = new MarkdownBuilder();

  const typeDeclaration = (model.type as any)
    ?.declaration as DeclarationReflection;

  md.add(SignatureTitleComponent(model));

  if (model.comment) {
    md.add(CommentsComponent(model.comment));
  }

  if (model.typeParameters?.length) {
    md.add(heading(nested ? 5 : 4, 'Type parameters'));
    md.add(TypeParameterTableComponent(model.typeParameters));
  }

  if (model.parameters?.length) {
    md.add(heading(nested ? 5 : 4, 'Parameters'));
    md.add(ParameterTable(model.parameters));
  }

  if (model.type && !model.parent?.kindOf(ReflectionKind.Constructor)) {
    if (model.type) {
      md.add(heading(nested ? 5 : 4, 'Returns'));
      md.add(TypeComponent(model.type, 'all'));
      if (model.comment?.returns) {
        md.add(comment(model.comment.returns));
      }
    }

    if (typeDeclaration?.signatures) {
      typeDeclaration.signatures.forEach((signature) => {
        md.add(SignatureComponent(signature, true));
      });
    }

    if (typeDeclaration?.children) {
      md.add(PropertyTableComponent(typeDeclaration.children));
    }
  }

  if (!nested) {
    md.add(SourcesComponent(model));
  }

  return md.print();
}
