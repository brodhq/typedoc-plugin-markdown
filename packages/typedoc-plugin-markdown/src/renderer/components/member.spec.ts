import { TestApp } from '../../../test/test-app';
import { MemberComponent } from './member';
import * as sourcesComponent from './sources';
describe(`Member Component:`, () => {
  let testApp: TestApp;

  beforeAll(async () => {
    jest
      .spyOn(sourcesComponent, 'SourcesComponent')
      .mockReturnValue('[SOURCES]');
    testApp = new TestApp(['declarations.ts']);
    await testApp.bootstrap();
  });

  test(`should compile a const with default value`, () => {
    const model = testApp.findReflectionByName('constWithDefaultValue');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile a let with default value`, () => {
    const model = testApp.findReflectionByName('letWithDefaultValue');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile an undefined declaration`, () => {
    const model = testApp.findReflectionByName('undefinedNumberDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile object literal declaration`, () => {
    const model = testApp.findReflectionByName('objectLiteralDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile object literal cast as a const`, () => {
    const model = testApp.findReflectionByName(
      'objectLiteralAsConstDeclaration',
    );
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile type literal declaration`, () => {
    const model = testApp.findReflectionByName('typeLiteralDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile declaration with double underscores in name and value`, () => {
    const model = testApp.findReflectionByName(
      '__DOUBLE_UNDERSCORES_DECLARATION__',
    );
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile any function type`, () => {
    const model = testApp.findReflectionByName('AnyFunctionType');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile function declaration`, () => {
    const model = testApp.findReflectionByName('functionDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile callable declaration`, () => {
    const model = testApp.findReflectionByName('callableDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile indexable declaration`, () => {
    const model = testApp.findReflectionByName('indexableDeclaration');
    expect(MemberComponent(model)).toMatchSnapshot();
  });

  test(`should compile enum delcaration`, () => {
    const model = testApp.findReflectionByName('EnumDeclarations');
    expect(MemberComponent(model)).toMatchSnapshot();
  });
});
