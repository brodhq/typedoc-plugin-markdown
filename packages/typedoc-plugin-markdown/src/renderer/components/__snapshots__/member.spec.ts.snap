// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member Component: should compile a const with default value 1`] = `
"### constWithDefaultValue

• \`Const\` **constWithDefaultValue**: \`\`\\"<foo>\\"\`\`

[SOURCES]"
`;

exports[`Member Component: should compile a let with default value 1`] = `
"### letWithDefaultValue

• \`Let\` **letWithDefaultValue**: \`string\` = \`'hello'\`

[SOURCES]"
`;

exports[`Member Component: should compile an undefined declaration 1`] = `
"### undefinedNumberDeclaration

• \`Let\` **undefinedNumberDeclaration**: \`number\`

[SOURCES]"
`;

exports[`Member Component: should compile any function type 1`] = `
"### AnyFunctionType

Ƭ **AnyFunctionType**<\`A\`\\\\>: (...\`input\`: \`any\`[]) => \`A\`

#### Type parameters


| Name | Type |
| :------ | :------ |
| \`A\` | \`A\` = \`any\` |


### Type declaration

▸ (...\`input\`): \`A\`


#### Parameters


| Name | Type |
| :------ | :------ |
| \`...input\` | \`any\`[] |


#### Returns

\`A\`

[SOURCES]

[SOURCES]"
`;

exports[`Member Component: should compile callable declaration 1`] = `
"### callableDeclaration

• \`Let\` **callableDeclaration**: \`Object\`

### Call signature

▸ (\`someArg\`): \`boolean\`


#### Parameters


| Name | Type |
| :------ | :------ |
| \`someArg\` | \`number\` |


#### Returns

\`boolean\`

[SOURCES]

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |
| \`arg2\` | \`number\` |


[SOURCES]"
`;

exports[`Member Component: should compile declaration with double underscores in name and value 1`] = `
"### \\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_

• \`Const\` **\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_**: typeof [\\\\_\\\\_DOUBLE\\\\_UNDERSCORES\\\\_DECLARATION\\\\_\\\\_](modules.md#__double_underscores_declaration__)

[SOURCES]"
`;

exports[`Member Component: should compile enum delcaration 1`] = `
"### EnumDeclarations

• **EnumDeclarations**: \`Object\`

[SOURCES]"
`;

exports[`Member Component: should compile function declaration 1`] = `
"### functionDeclaration

▸ \`Let\` **functionDeclaration**(\`someArg\`): \`boolean\`


#### Parameters


| Name | Type |
| :------ | :------ |
| \`someArg\` | \`number\` |


#### Returns

\`boolean\`

[SOURCES]"
`;

exports[`Member Component: should compile indexable declaration 1`] = `
"### indexableDeclaration

• \`Let\` **indexableDeclaration**: \`Object\`

#### Index signature

▪ []: \`string\`

#### Type declaration


| Name | Type |
| :------ | :------ |
| \`arg1\` | \`string\` |


[SOURCES]"
`;

exports[`Member Component: should compile object literal cast as a const 1`] = `
"### objectLiteralAsConstDeclaration

• \`Const\` **objectLiteralAsConstDeclaration**: \`Object\`

Comments

#### Type declaration


| Name | Type | Description |
| :------ | :------ | :------ |
| \`Prop1\` | \`\`\\"Prop1\\"\`\` | Comment for Prop1. |
| \`Prop2\` | \`\`\\"Prop2\\"\`\` | Comment for Prop2. |
| \`Prop3\` | \`\`\\"Prop3\\"\`\` | - |


[SOURCES]"
`;

exports[`Member Component: should compile object literal declaration 1`] = `
"### objectLiteralDeclaration

• \`Const\` **objectLiteralDeclaration**: \`Object\`

#### Type declaration


| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA\` | \`number\` | - |
| \`valueB\` | \`boolean\` | - |
| \`valueC\` | \`Object\` | - |
| \`valueX\` | \`Object\` | description for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueZ\` | \`string\` | - |
| \`valueY\` | () => \`string\` | description for valueY |
| \`valueZ\` | \`string\` | description for valueZ |


[SOURCES]"
`;

exports[`Member Component: should compile type literal declaration 1`] = `
"### typeLiteralDeclaration

• \`Let\` **typeLiteralDeclaration**: \`Object\`

#### Type declaration


| Name | Type | Description |
| :------ | :------ | :------ |
| \`valueA?\` | \`number\` | Comments for valueA |
| \`valueB?\` | \`boolean\` | - |
| \`valueX\` | \`Object\` | Comment for valueX |
| \`valueX.valueA\` | \`number\`[] | - |
| \`valueX.valueY\` | (\`z\`: \`string\`) => { \`a\`: \`string\` ; \`b\`: \`string\`  } | - |
| \`valueX.valueZ\` | \`string\` | Nested comment for valueZ |
| \`valueY\` | () => \`string\` | - |
| \`valueZ\` | \`string\` | Comment for valueZ |


[SOURCES]"
`;
